tuna/gen_fastdb.py

```diff
@@ -0,0 +1,298 @@
+"""generate a fastdb dataframe from a finddb dataframe
+
+finddb contains convolution problems and the solvers that solve them.
+finddb may list multiple solvers per convolution problem, some solving the
+problem faster than the others. fastdb is generated from finddb by looking
+at each convolution problem in finddb and keeping only the fastest solver
+for it.
+"""
+
+import os
+import argparse
+
+from tuna.utils import logging
+import tuna.utils.tools.df as df_tools
+import tuna.utils.tools.io as io_tools
+from tuna.gen_finddb import load_finddb
+from tuna.utils.db_utility import get_id_solvers
+from tuna.utils.ANSI_formatting import ANSIColors
+from tuna.utils.finddb_like_utils import get_solver_counts
+from tuna.utils.helpers import print_heading, filter_out, map_list, \
+  is_substr, invert_dict, sort_dict, as_heading
+
+_DEFAULT_INPUT_DIR = os.path.join(os.getcwd(), 'finddb_')
+_DEFAULT_OUTPUT_DIR = os.path.join(os.getcwd(), 'finddb_')
+
+_, _ID_TO_SOLVER = get_id_solvers()
+_SOLVER_TO_ID = invert_dict(_ID_TO_SOLVER)
+_DEFAULT_COLS_WITH_CONV_PARAMS = ['fdb_key']
+
+
+# pylint: disable-next=too-many-locals, too-many-branches
+def check_finddb(finddb,
+                 cols_with_conv_params=None,
+                 strict=False,
+                 verbosity=0,
+                 tag=''):
+  """check finddb and raise erros or log warnings"""
+  if cols_with_conv_params is None:
+    cols_with_conv_params = _DEFAULT_COLS_WITH_CONV_PARAMS
+
+  # pylint: disable-next=invalid-name
+  NO_ENTRIES, INVALID_ENTRIES, OPENCL_ON, MULTIPLE_SESSIONS, DUPLICATE_ENTRIES = 0, 1, 2, 3, 7
+  issues = []
+
+  logging.log(f'checking {tag}finddb...', end_char='\r')
+
+  if len(finddb) == 0:
+    if strict:
+      raise AssertionError(f'{tag}finddb contains no entry')
+    logging.error(f'{tag}finddb contains no entry')
+    issues.append(NO_ENTRIES)
+
+  if (finddb['valid'] != 1).any():
```

Comment:
this should be covered by your db selection query, which should always select based on the filter valid=true. This makes sure that you never pick up invalid entries to look at. Then this check here, can go away

Reply:
added a flag called `invalid too`, so now you have to explicitly specify if you want invalid entries